### ISSUE TITLE
NaNs in KL Loss

### ISSUE THREAD

**Issue by Bleyddyn** [AppliedDataSciencePartners/WorldModels#9]
I'm getting NaN's in the KL Loss (specifically the K.square(vae_z_mean), part) in the first epoch of the first batch when trying to train the vae. First the KL loss (and overall loss) blows up, then starts going back down, but turns into NaN's before long:

13696/48000 [=======>......................] - ETA: 4:32 - loss: 289661273.9050 - vae_r_loss: 0.1846 - vae_kl_loss: 289661273.7244
23904/48000 [=============>................] - ETA: 3:13 - loss: 191863659.1293 - vae_r_loss: 0.1638 - vae_kl_loss: 191863658.9721
24992/48000 [==============>...............] - ETA: 3:04 - loss: nan - vae_r_loss: nan - vae_kl_loss: nan                          

Mac OS X 10.13
Python 3.6.2
tf.version '1.7.0'
keras.version '2.1.5'
gym.version '0.10.5'
np.__version__ '1.14.2'

I've tried over from the beginning a couple of times, including a clean virtualenv. I've looked at some of the stats from the data files, and don't see anything too suspicious there.

Any suggestions would be welcome.

Changes I've tried:

* SGD instead of RMSProp
* Normalizing images (-mean, /std)
* Increasing and decreasing batch size
* Changing the constant in vae_r_loss() [There was some improvement when it was 1.0, more testing needed]
* Train on a friend's generated data, which he used to successfully train a vae (on a GPU).

I haven't tried running on a GPU since I don't have a supported one.

I have successfully run steps 1-4 on my work computer (which i shouldn't be doing) so no NaN's for:

Mac OS X 10.9.5
CPU
Python 3.6.2
tf.version '1.3.0'
keras.version '2.1.3'
gym.version '0.10.4'
np.__version__ '1.14.2'

I'm currently trying to run step 5 on my home machine using saved weights from work. Not a real solution, though.

**Comment by davidADSP** [comment]
Yes this is a strange one - I also found that I get NaNs in the loss when the machine isn't powerful enough. The only fix that I know of is to run on a machine with more CPU / RAM. If anyone can provide an explanation as to why this is the case I'd be interested to hear it!

**Comment by Bleyddyn** [comment]
@davidADSP  Very bizarre, but I agree that fits with what I've seen. I've never heard or read of anything like that, though, so no idea how fix it other than buy a new machine.

**Comment by Bleyddyn** [comment]
I've made enough progress on this that I'm going to close it.

I added a relu activation to the vae_z_log_var layer and I changed the constant in the vae_r_loss method to 0.5. Both of those changes seem to be necessary. I can then train on at least 100k images and the results when I sample from the vae look at least somewhat reasonable. Although, they all look very, very similar with nothing but a straight stretch of track, so I'm not sure how well the rest of the model is going to work.

Both changes are in the vae/arch.py part of this commit: https://github.com/Bleyddyn/WorldModels/commit/056f6636c9e3ae5e6eff3680f108d4a54371f2f2
